README.md
---
title: "Package moonBook"
author: "Keon-Woong Moon"
date: "2017-11-11"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{package moonBook}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---


## Function "mytable"

Function "mytable"" produce table for descriptive analysis easily. It is most useful to make table to describe baseline characteristics common in medical research papers. 

## Basic Usage


```r
require(moonBook)
data(acs)
mytable(Dx~.,data=acs)
```

```

                 Descriptive Statistics by 'Dx'                 
_________________________________________________________________ 
                     NSTEMI       STEMI     Unstable Angina   p  
                    (N=153)      (N=304)        (N=400)    
----------------------------------------------------------------- 
 age              64.3 ± 12.3  62.1 ± 12.1    63.8 ± 11.0   0.073
 sex                                                        0.012
   - Female        50 (32.7%)   84 (27.6%)    153 (38.2%)        
   - Male         103 (67.3%)  220 (72.4%)    247 (61.8%)        
 cardiogenicShock                                           0.000
   - No           149 (97.4%)  256 (84.2%)   400 (100.0%)        
   - Yes           4 ( 2.6%)    48 (15.8%)     0 ( 0.0%)         
 entry                                                      0.001
   - Femoral       58 (37.9%)  133 (43.8%)    121 (30.2%)        
   - Radial        95 (62.1%)  171 (56.2%)    279 (69.8%)        
 EF               55.0 ±  9.3  52.4 ±  9.5    59.2 ±  8.7   0.000
 height           163.3 ±  8.2 165.1 ±  8.2  161.7 ±  9.7   0.000
 weight           64.3 ± 10.2  65.7 ± 11.6    64.5 ± 11.6   0.361
 BMI              24.1 ±  3.2  24.0 ±  3.3    24.6 ±  3.4   0.064
 obesity                                                    0.186
   - No           106 (69.3%)  209 (68.8%)    252 (63.0%)        
   - Yes           47 (30.7%)   95 (31.2%)    148 (37.0%)        
 TC               193.7 ± 53.6 183.2 ± 43.4  183.5 ± 48.3   0.057
 LDLC             126.1 ± 44.7 116.7 ± 39.5  112.9 ± 40.4   0.004
 HDLC             38.9 ± 11.9  38.5 ± 11.0    37.8 ± 10.9   0.501
 TG               130.1 ± 88.5 106.5 ± 72.0  137.4 ± 101.6  0.000
 DM                                                         0.209
   - No            96 (62.7%)  208 (68.4%)    249 (62.2%)        
   - Yes           57 (37.3%)   96 (31.6%)    151 (37.8%)        
 HBP                                                        0.002
   - No            62 (40.5%)  150 (49.3%)    144 (36.0%)        
   - Yes           91 (59.5%)  154 (50.7%)    256 (64.0%)        
 smoking                                                    0.000
   - Ex-smoker     42 (27.5%)   66 (21.7%)    96 (24.0%)         
   - Never         50 (32.7%)   97 (31.9%)    185 (46.2%)        
   - Smoker        61 (39.9%)  141 (46.4%)    119 (29.8%)        
----------------------------------------------------------------- 
```

The first argument of function `mytable` is an object of class `formula`. Left side of ~ must contain the name of one grouping variable or two grouping variables in an additive way(e.g. sex+group~), and the right side of ~ must have variables in an additive way. `.` is allowed on the right side of formula which means all variables in the data.frame specified by the 2nd argument `data`. The sample data 'acs' containing demographic data and laboratory data of 857 patients with acute coronary syndrome(ACS). For more information about the data acs, type ?acs in your R console.  


```r
str(acs)
```

```
'data.frame':	857 obs. of  17 variables:
 $ age             : int  62 78 76 89 56 73 58 62 59 71 ...
 $ sex             : chr  "Male" "Female" "Female" "Female" ...
 $ cardiogenicShock: chr  "No" "No" "Yes" "No" ...
 $ entry           : chr  "Femoral" "Femoral" "Femoral" "Femoral" ...
 $ Dx              : chr  "STEMI" "STEMI" "STEMI" "STEMI" ...
 $ EF              : num  18 18.4 20 21.8 21.8 22 24.7 26.6 28.5 31.1 ...
 $ height          : num  168 148 NA 165 162 153 167 160 152 168 ...
 $ weight          : num  72 48 NA 50 64 59 78 50 67 60 ...
 $ BMI             : num  25.5 21.9 NA 18.4 24.4 ...
 $ obesity         : chr  "Yes" "No" "No" "No" ...
 $ TC              : num  215 NA NA 121 195 184 161 136 239 169 ...
 $ LDLC            : int  154 NA NA 73 151 112 91 88 161 88 ...
 $ HDLC            : int  35 NA NA 20 36 38 34 33 34 54 ...
 $ TG              : int  155 166 NA 89 63 137 196 30 118 141 ...
 $ DM              : chr  "Yes" "No" "No" "No" ...
 $ HBP             : chr  "No" "Yes" "Yes" "No" ...
 $ smoking         : chr  "Smoker" "Never" "Never" "Never" ...
```


## Choosing grouping variable(s) and row-variable(s)

You can choose the grouping variable(s) and row-variable(s) with the `formula`. 


```r
mytable(sex~age+Dx,data=acs)
```

```

         Descriptive Statistics by 'sex'         
__________________________________________________ 
                       Female       Male       p  
                       (N=287)     (N=570)  
-------------------------------------------------- 
 age                 68.7 ± 10.7 60.6 ± 11.2 0.000
 Dx                                          0.012
   - NSTEMI          50 (17.4%)  103 (18.1%)      
   - STEMI           84 (29.3%)  220 (38.6%)      
   - Unstable Angina 153 (53.3%) 247 (43.3%)      
-------------------------------------------------- 
```

You can choose row-variable(s) with `.` and `+` and `-` and variable name in an additive way.


```r
mytable(am~.-hp-disp-cyl-carb-gear,data=mtcars)
```

```

   Descriptive Statistics by 'am'  
____________________________________ 
            0           1        p  
         (N=19)      (N=13)   
------------------------------------ 
 mpg   17.1 ±  3.8 24.4 ±  6.2 0.000
 drat   3.3 ±  0.4  4.0 ±  0.4 0.000
 wt     3.8 ±  0.8  2.4 ±  0.6 0.000
 qsec  18.2 ±  1.8 17.4 ±  1.8 0.206
 vs                            0.556
   - 0 12 (63.2%)   6 (46.2%)       
   - 1  7 (36.8%)   7 (53.8%)       
------------------------------------ 
```

## Method for continuous variables

By default continuous variables are analyzed as normal-distributed and are described with mean and standard deviation. To change default options, you can use the `method` argument. Possible values of `method` argument are:

- 1: forces analysis as normal-distributed, default value
- 2: forces analysis as continuous non-normal
- 3: performs a Shapiro-Wilks test to decide between normal or non-normal

When continuous variables are analyzed as non-normal, they are described with median and interquartile range.


```r
mytable(sex~height+weight+BMI,data=acs,method=3)
```

```

           Descriptive Statistics by 'sex'          
_____________________________________________________ 
              Female               Male           p  
              (N=287)             (N=570)      
----------------------------------------------------- 
 height 155.0 [150.0;158.0] 168.0 [164.0;172.0] 0.000
 weight  58.0 [50.0;63.0]    68.0 [62.0;75.0]   0.000
 BMI     24.0 [22.1;26.2]    24.2 [22.2;26.2]   0.471
----------------------------------------------------- 
```

Because the `method` argument is selected as 3, a Shapiro-Wilk test normality test is used to decide if the variable is normal or non-normal distributed. Note that `height` and `BMI` was described as mean $\pm$ sd, whereas the weight was described as median and interquartile range. 

## choice of variable : categorical or continuous variable - my way

In many cases, categorical variables are usually coded as numeric. For example, many people usually code 0 and 1 instead of "No" and "Yes". Similarly, factor variables with three or four levels are coded 0/1/2 or 0/1/2/3. In many cases, if we analyze these variables as continuous variables, we are not able to get the right result. In `mytable`, variables with less than five `unique` values are treated as a categorical variables. 


```r
mytable(am~.,data=mtcars)
```

```

    Descriptive Statistics by 'am'    
_______________________________________ 
             0            1         p  
          (N=19)        (N=13)   
--------------------------------------- 
 mpg    17.1 ±  3.8  24.4 ±  6.2  0.000
 cyl                              0.013
   - 4   3 (15.8%)    8 (61.5%)        
   - 6   4 (21.1%)    3 (23.1%)        
   - 8  12 (63.2%)    2 (15.4%)        
 disp  290.4 ± 110.2 143.5 ± 87.2 0.000
 hp    160.3 ± 53.9  126.8 ± 84.1 0.180
 drat    3.3 ±  0.4   4.0 ±  0.4  0.000
 wt      3.8 ±  0.8   2.4 ±  0.6  0.000
 qsec   18.2 ±  1.8  17.4 ±  1.8  0.206
 vs                               0.556
   - 0  12 (63.2%)    6 (46.2%)        
   - 1   7 (36.8%)    7 (53.8%)        
 gear                             0.000
   - 3  15 (78.9%)     0 ( 0.0%)       
   - 4   4 (21.1%)    8 (61.5%)        
   - 5   0 ( 0.0%)    5 (38.5%)        
 carb    2.7 ±  1.1   2.9 ±  2.2  0.781
--------------------------------------- 
```

In `mtcars` data, all variables are expressed as numeric. But as you can see, `cyl`, `vs` and `gear` is treated as categorical variables. The `carb` variables has six `unique` values and treated as continuous variables. If you wanted the `carb` variable to be treated as categorical variable, you can changed the `max.ylev` argument.


```r
mytable(am~carb,data=mtcars,max.ylev=6)
```

```

  Descriptive Statistics by 'am' 
__________________________________ 
           0          1        p  
         (N=19)     (N=13)  
---------------------------------- 
 carb                        0.284
   - 1 3 (15.8%)  4 (30.8%)       
   - 2 6 (31.6%)  4 (30.8%)       
   - 3 3 (15.8%)   0 ( 0.0%)      
   - 4 7 (36.8%)  3 (23.1%)       
   - 6  0 ( 0.0%) 1 ( 7.7%)       
   - 8  0 ( 0.0%) 1 ( 7.7%)       
---------------------------------- 
```

## Combining tables

If you wanted to make two separate tables and combine into one table, **`mytable` is the function of choice**. For example, if you wanted to build separate table for female and male patients stratified by presence or absence of DM and combine it, 


```r
mytable(sex+DM~.,data=acs)
```

```

                 Descriptive Statistics stratified by 'sex' and 'DM'                
_____________________________________________________________________________________ 
                                    Male                             Female             
                     -------------------------------- ------------------------------- 
                          No           Yes        p        No          Yes        p  
                       (N=380)       (N=190)           (N=173)      (N=114)        
------------------------------------------------------------------------------------- 
 age                 60.9 ± 11.5   60.1 ± 10.6  0.472 69.3 ± 11.4  67.8 ±  9.7  0.257
 cardiogenicShock                               0.685                           0.296
   - No              355 (93.4%)   175 (92.1%)        168 (97.1%)  107 (93.9%)       
   - Yes              25 ( 6.6%)   15 ( 7.9%)          5 ( 2.9%)    7 ( 6.1%)        
 entry                                          0.552                           0.665
   - Femoral         125 (32.9%)   68 (35.8%)          74 (42.8%)   45 (39.5%)       
   - Radial          255 (67.1%)   122 (64.2%)         99 (57.2%)   69 (60.5%)       
 Dx                                             0.219                           0.240
   - NSTEMI           71 (18.7%)   32 (16.8%)          25 (14.5%)   25 (21.9%)       
   - STEMI           154 (40.5%)   66 (34.7%)          54 (31.2%)   30 (26.3%)       
   - Unstable Angina 155 (40.8%)   92 (48.4%)          94 (54.3%)   59 (51.8%)       
 EF                  56.5 ±  8.3   53.9 ± 11.0  0.007 56.0 ± 10.1  56.6 ± 10.0  0.655
 height              168.1 ±  5.8 167.5 ±  6.7  0.386 153.9 ±  6.5 153.6 ±  5.8 0.707
 weight              68.1 ± 10.4   69.8 ± 10.2  0.070 56.5 ±  8.7  58.4 ± 10.0  0.106
 BMI                 24.0 ±  3.1   24.9 ±  3.5  0.005 23.8 ±  3.2  24.8 ±  4.0  0.046
 obesity                                        0.027                           0.359
   - No              261 (68.7%)   112 (58.9%)        121 (69.9%)   73 (64.0%)       
   - Yes             119 (31.3%)   78 (41.1%)          52 (30.1%)   41 (36.0%)       
 TC                  184.1 ± 46.7 181.8 ± 44.5  0.572 186.0 ± 43.1 193.3 ± 60.8 0.274
 LDLC                117.9 ± 41.8 112.1 ± 39.4  0.115 116.3 ± 35.2 119.8 ± 48.6 0.519
 HDLC                38.4 ± 11.4   36.8 ±  9.6  0.083 39.2 ± 10.9  38.8 ± 12.2  0.821
 TG                  115.2 ± 72.2 153.4 ± 130.7 0.000 114.2 ± 82.4 128.4 ± 65.5 0.112
 HBP                                            0.000                           0.356
   - No              205 (53.9%)   68 (35.8%)          54 (31.2%)   29 (25.4%)       
   - Yes             175 (46.1%)   122 (64.2%)        119 (68.8%)   85 (74.6%)       
 smoking                                        0.386                           0.093
   - Ex-smoker       101 (26.6%)   54 (28.4%)          34 (19.7%)   15 (13.2%)       
   - Never            77 (20.3%)   46 (24.2%)         118 (68.2%)   91 (79.8%)       
   - Smoker          202 (53.2%)   90 (47.4%)          21 (12.1%)   8 ( 7.0%)        
------------------------------------------------------------------------------------- 
```

## For more beautiful output : myhtml

If you want more beautiful table in your R markdown file, you can use myhtml function.


```r
out=mytable(Dx~.,data=acs)
myhtml(out)
```

<head><style>
        table, th, td {
        border: 1px solid #bcbcbc;
    } </style></head>
<table cellpadding=10 cellspacing=5><caption>Descriptive Statistics by 'Dx'</caption><tr>
<th>Dx</th><th>NSTEMI<br/>(N=153)</th><th>STEMI<br/>(N=304)</th><th>Unstable Angina<br/>(N=400)</th><th>p</th></tr>
<tr><td>age             </td><td>64.3 ± 12.3</td><td>62.1 ± 12.1</td><td>63.8 ± 11.0</td><td>0.073</td></tr>
<tr><td>sex             </td><td></td><td></td><td></td><td>0.012</td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Female      </td><td>50 (32.7%)</td><td>84 (27.6%)</td><td>153 (38.2%)</td><td></td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Male        </td><td>103 (67.3%)</td><td>220 (72.4%)</td><td>247 (61.8%)</td><td></td></tr>
<tr><td>cardiogenicShock</td><td></td><td></td><td></td><td>0.000</td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; No          </td><td>149 (97.4%)</td><td>256 (84.2%)</td><td>400 (100.0%)</td><td></td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Yes         </td><td>4 ( 2.6%)</td><td>48 (15.8%)</td><td> 0 ( 0.0%)</td><td></td></tr>
<tr><td>entry           </td><td></td><td></td><td></td><td>0.001</td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Femoral     </td><td>58 (37.9%)</td><td>133 (43.8%)</td><td>121 (30.2%)</td><td></td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Radial      </td><td>95 (62.1%)</td><td>171 (56.2%)</td><td>279 (69.8%)</td><td></td></tr>
<tr><td>EF              </td><td>55.0 ±  9.3</td><td>52.4 ±  9.5</td><td>59.2 ±  8.7</td><td>0.000</td></tr>
<tr><td>height          </td><td>163.3 ±  8.2</td><td>165.1 ±  8.2</td><td>161.7 ±  9.7</td><td>0.000</td></tr>
<tr><td>weight          </td><td>64.3 ± 10.2</td><td>65.7 ± 11.6</td><td>64.5 ± 11.6</td><td>0.361</td></tr>
<tr><td>BMI             </td><td>24.1 ±  3.2</td><td>24.0 ±  3.3</td><td>24.6 ±  3.4</td><td>0.064</td></tr>
<tr><td>obesity         </td><td></td><td></td><td></td><td>0.186</td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; No          </td><td>106 (69.3%)</td><td>209 (68.8%)</td><td>252 (63.0%)</td><td></td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Yes         </td><td>47 (30.7%)</td><td>95 (31.2%)</td><td>148 (37.0%)</td><td></td></tr>
<tr><td>TC              </td><td>193.7 ± 53.6</td><td>183.2 ± 43.4</td><td>183.5 ± 48.3</td><td>0.057</td></tr>
<tr><td>LDLC            </td><td>126.1 ± 44.7</td><td>116.7 ± 39.5</td><td>112.9 ± 40.4</td><td>0.004</td></tr>
<tr><td>HDLC            </td><td>38.9 ± 11.9</td><td>38.5 ± 11.0</td><td>37.8 ± 10.9</td><td>0.501</td></tr>
<tr><td>TG              </td><td>130.1 ± 88.5</td><td>106.5 ± 72.0</td><td>137.4 ± 101.6</td><td>0.000</td></tr>
<tr><td>DM              </td><td></td><td></td><td></td><td>0.209</td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; No          </td><td>96 (62.7%)</td><td>208 (68.4%)</td><td>249 (62.2%)</td><td></td></tr>
<tr><td> &nbsp;&nbsp;&nbsp; Yes         </td><td>57 (37.3%)</td><td>96 (31.6%)</td><td>151 (37.8%)</td><td></td></tr>
<tr><td>HBP             </td><td></td><td></td><td></td><td>0.002</td></tr>

NEWS
# moonBook 0.3.4
==============
(2023-Nov-28)

* new argument 'useSE' in mytable function to use standard error instead of standard deviation

# moonBook 0.3.3
==============
(2023-Mar-12)

* new argument 'clean_names' in mytable function for clean column names
* bug-fixed


# moonBook 0.3.2
==============
(2022-Nov-16)

* add function mycsv.mytable.df

# moonBook 0.3.1
==============
(2021-Dec-29)

* bugs fix in function extractOR and extractHR


# moonBook 0.3.0
==============
(2021-Dec-26)

* bugs fix in function extractOR and extractHR

# moonBook 0.2.9
==============
(2021-Dec-20)

* new argument 'missing' in mytable function for missing data analysis
* mytable() print red character in R console when p < 0.05


# moonBook 0.2.8
==============
(2021-Sep-4)

* Bug fixed :


# moonBook 0.2.7
==============
(2021-July-7)

* Bug fixed :


# moonBook 0.2.6
==============
(2021-Apr-27)

* Bug fixed :

# moonBook 0.2.5
==============
(2021-Jan-5)

* Bug fixed :

# moonBook 0.2.4
==============
(2021-Jan-5)

* Bug fixed

# moonBook 0.2.3
==============
(2018-July-21)

* Bug fixed


# moonBook 0.2.2
==============
(2018-June-18)

* New function getMapping() added

* new argument 'mapping' added to function addLabelDf()

# moonBook 0.2.1
==============
(2018-May-30)

* New argument 'catMethod' added to mytable() functions to select method for categorical variables

* New function deleteRows() added

# moonBook 0.2.0
==============
(2018-May-26)

* New function compress() and addComma() added

* New vignette "moonBook_update" added

# moonBook 0.1.9
==============
(2018-May-16)

* New S3 method mytable.formula and mytable.data.frame added

* mytable() functions can handle column label and value labels

# moonBook 0.1.8
===============
(2018-Apr-25)

* Bug fixed


# moonBook 0.1.7
==============
(2016-Nov-23)

* Bug fixed in function mytable()


# moonBook 0.1.6
==============
(2016-Nov-11)

* Bug fixed

* default method for compute confidence interval for glm model changed from confint to confint.default


# moonBook 0.1.4
==============
(2015-Jul-07)

* Bug fixed

* "mytable" function can show the total value when show.total parameter is TRUE.


# moonBook 0.1.3
==============
(2015-Feb-14)

* Bug fixed

* "myhtml" function makes a better-looking table

# moonBook 0.1.2
==============
(2014-Dec-7)

* new S3 method for class "data.frame" added to function "mylatex" and "myhtml"

* more option to function "mylatex" to adjust caption placement and position


# moonBook 0.1.1
==============
(2014-Dec-5)

* new function "mycsv" makes a csv file with an object of class "cbind.mytable" or "mytable" made by function "mytable"

* new function "myhtml" makes a html table with an object of class "cbind.mytable" or "mytable" made by function "mytable"

* Bug fixed



# moonBook 0.1.0
==============

* new function "ORplot" which draw odds ratio plot with an object of class glm

* new function "HRplot" which draw hazard ratio plot with an object of class coxph or a resultant data.frame of function "mycph"

* new function "mycph" which perform coxph of individual expecting variables from formula parameter

* new function "mytable" which automatically produce a table for descriptive
  analysis easily. It is most useful to make a table to describe baseline
  charateristics common in medical research papers. An object of class
  "cbind.mytable" or "mytable" will be made by this function

* new function "mylatex" makes a latex table with an object of class "cbind.mytable" or "mytable" made by function "mytable"

* new function "densityplot" which draw Kernel density plot from formula

* new function "rank2group" which rank a numeric vector and returns a new ordinal vector

* new data "radial" added

* new data "acs" added


R/mylatex.R
#'Demographic data of 857 patients with ACS
#'
#'A dataset containing demographic data and laboratory data of 857 patients with
#'acute coronary syndrome(ACS).
#'
#'
#'@format A data frame with 857 rows and 17 variables:
#'\describe{
#'  \item{age}{patient age in years}
#'  \item{sex}{"Male" or "Female"}
#'  \item{cardiogenicShock}{"No" or "Yes"}
#'  \item{entry}{vascular access route, either "Femoral" or "Radial"}
#'  \item{Dx}{Final diagnosis, One of the followings : STEMI, NSTEMI or Unstable Angina}
#'  \item{EF}{ejection fraction, percentage by echocardiography}
#'  \item{height}{height in centimeter}
#'  \item{weight}{weight in kilogram}
#'  \item{BMI}{body mass index in kg/m2}
#'  \item{obesity}{obesity, "No" or "Yes"}
#'  \item{TC}{total cholesterol level in mg/dL}
#'  \item{LDLC}{low density lipoprotein cholesterol level in mg/dL}
#'  \item{HDLC}{high density lipoprotein cholesterol level in mg/dL}
#'  \item{TG}{triglyceride level in mg/dL}
#'  \item{DM}{history of diabetes mellitus,"No" or "Yes"}
#'  \item{HBP}{history of hypertension,"No" or "Yes"}
#'  \item{smoking}{history of smoking, One of the followings : "Never","Ex-smoker","Smoker"}
#'}
#'@name acs
NULL

#'Demographic data of 115 patients performing IVUS(intravascular ultrasound)
#'examination of a radial artery.
#'
#'A dataset containing demographic data and laboratory data of 115 patients performing IVUS(intravascular ultrasound)
#'examination of a radial artery after tansradial coronary angiography.
#'
#'
#'@format A data frame with 115 rows and 15 variables:
#'\describe{
#'  \item{male}{if Male, 1; if Female 0}
#'  \item{age}{patient age in years}
#'  \item{height}{height in centimeter}
#'  \item{weight}{weight in kilogram}
#'  \item{HBP}{history of hypertension, 1 for yes or 0 for no}
#'  \item{DM}{history of diabetes mellitus, 1 for yes or 0 for no}
#'  \item{smoking}{history of smoking, One of the followings : "non-smoker","ex-smoker","smoker"}
#'  \item{TC}{total cholesterol level in mg/dL}
#'  \item{TG}{triglyceride level in mg/dL}
#'  \item{HDL}{high density lipoprotein cholesterol level in mg/dL}
#'  \item{LDL}{low density lipoprotein cholesterol level in mg/dL}
#'  \item{hsCRP}{high-sensitive C reactive protein}
#'  \item{NTAV}{normalized total atheroma volume measured by IVUS in cubic mm}
#'  \item{PAV}{percent atheroma volume in percentage}
#'  \item{sex}{Factor with two levels; "Male" or "Female"}
#'}
#'@name radial
NULL




#' Exporting "cbind.mytable","mytable" to LaTeX format
#'
#' Exporting "cbind.mytable","mytable" to LaTeX format
#'@param myobj An object of class 'mytable'
#'@param size An integer indicating font size, defaulting is 5.
#'@param caption A character
#'@param caption.placement The caption will be have placed at the top of the table
#'        if caption.placement is "top" and at the bottom of the table
#'        if it equals "bottom". Default value is "top".
#'@param caption.position The caption will be have placed at the center of the table
#'        if caption.position is "center" or "c", and at the left side of the table
#'        if it equals "left" or "l", and at the right side of the table
#'        if it equals "right" or "r". Default value is "center".
#' @export
#'@examples
#' require(moonBook)
#' out=mytable(sex~.,data=acs)
#' mylatex(out)
#' out1=mytable(sex+Dx~.,data=acs)
#' mylatex(out1,size=6)
mylatex=function(myobj,size=5,caption=NULL,caption.placement="top",
                 caption.position="c") UseMethod("mylatex")


#'@describeIn mylatex Exporting "cbind.mytable","mytable" to LaTeX format
#' @export
mylatex.default=function(myobj,size=5,caption=NULL,caption.placement="top",
                         caption.position="c") {

    cat("mylatex function only applicable to data.frame, mytable or cbind.mytable\n")
}


#'@describeIn mylatex Exporting "cbind.mytable","mytable" to LaTeX format
#' @export
mylatex.mytable=function(myobj,size=5,caption=NULL,caption.placement="top",
                         caption.position="c") {

    ## Generate latex table for class mytable
    result=obj2linecount(myobj)
    y=result$y
    out1=result$out1
    cn=result$cn
    ncount=result$ncount
    col.length=result$col.length
    linelength=result$linelength

    if(identical(caption.placement,"bottom") | identical(caption.placement,"b"))
            caption.placement="bottom"
    else caption.placement="top"
    if(identical(caption.position,"left")|identical(caption.position,"l"))
            caption.position="l"
    else if(identical(caption.position,"right")|identical(caption.position,"r"))
            caption.position="r"
    else caption.position="c"

    if(!is.numeric(size)) size=5
    else if(size<0 | size>10) size=5
    Fontsize=c("tiny","scriptsize","footnotesize","small","normalsize",
               "large","Large","LARGE","huge","Huge")
    cat("\\begin{table}[!hbp]\n")

    cat(paste("\\begin{",Fontsize[size],"}\n",sep=""))
    head=c("\\begin{tabular}{l")
    for(i in 2 : length(cn)) { head=paste(head,"c",sep="")}
    head=paste(head,"}\n",sep="")
    cat(head)
    if(is.null(caption)) caption= paste("Descriptive Statistics by ",y,sep="")
    if(caption.placement=="top")
        cat(paste("\\multicolumn{",length(cn),"}{",
                  caption.position,"}{",caption,"}\\\\ \n",sep=""))
    cat("\\hline\n")
    firstrow=cn[1]
    for(i in 2:length(ncount)) { firstrow=paste(firstrow,cn[i],sep=" & ")}
    for(i in 1:(length(cn)-length(ncount))){
        firstrow=paste(firstrow," & \\multirow{2}{*}{",
                       cn[length(ncount)+i],"}",sep="")
    }
    cat(paste(firstrow,"\\\\ \n",sep=""))

    secondrow=ncount[1]
    for(i in 2:length(ncount)) {
        secondrow=paste(secondrow,ncount[i],sep=" & ")
    }
    for(i in 1:(length(cn)-length(ncount)))
        secondrow=paste(secondrow," & ",sep="")
    cat(paste(secondrow," \\\\ \n",sep=""))
    cat("\\hline\n")

    for(i in 1:dim(out1)[1]){
        temp=r(out1[i,1])
        for(j in 2:(length(cn))) {
            temp=paste(temp,r(out1[i,j]),sep=" & ")
        }
        cat(paste(temp,"\\\\ \n",sep=""))
    }
    cat("\\hline\n")
    if(caption.placement=="bottom")
        cat(paste("\\multicolumn{",length(cn),"}{",
                  caption.position,"}{",caption,"}\\\\ \n",sep=""))
    cat("\\end{tabular}\n")
    cat(paste("\\end{",Fontsize[size],"}\n",sep=""))
    cat("\\end{table}\n")
}

#'Subfunction used in mylatex
#'
#' @param string a character vector
r=function(string) {
    string=gsub("%","\\%",string,fixed=TRUE)
    string
}

#'@describeIn mylatex Exporting "cbind.mytable","mytable" to LaTeX format
#' @export
mylatex.cbind.mytable=function(myobj,size=5,caption=NULL,
                               caption.placement="top",caption.position="c"){
    ## Generate latex table for cbind.mytable

    tcount=length(myobj) # number of tables
    tnames=unlist(attr(myobj,"caption"))
    group=attr(myobj,"group")
    result=list()

    for(i in 1:tcount) result[[i]]=obj2linecount(myobj[[i]])
    if(!is.numeric(size)) size=5
    else if(size<0 | size>10) size=5

    Fontsize=c("tiny","scriptsize","footnotesize","small","normalsize",
               "large","Large","LARGE","huge","Huge")
    cat("\\begin{table}[!hbp]\n")
    cat(paste("\\begin{",Fontsize[size],"}\n",sep=""))
    cat("\\centering\n")
    if(is.null(caption)) {
        caption=paste("Descriptive Statistics stratified by ",group[1],sep="")
        for(i in 2:tcount) caption=paste(caption," and ",group[i],sep="")
    }

    if(identical(caption.placement,"bottom") | identical(caption.placement,"b"))
        caption.placement="bottom"
    else caption.placement="top"
    if(identical(caption.position,"left")|identical(caption.position,"l"))
        caption.position="l"
    else if(identical(caption.position,"right")|identical(caption.position,"r"))
        caption.position="r"
    else caption.position="c"

    colno=length(result[[1]]$cn)
    # number of total column
    tcn=colno*tcount

    head=c("\\begin{tabular}{l")
    for(i in 2 : tcn) { head=paste(head,"c",sep="")}
    head=paste(head,"}\n",sep="")
    cat(head)
    if(caption.placement=="top")
        cat(paste("\\multicolumn{",tcn,"}{",
                  caption.position,"}{",caption,"}\\\\ \n",sep=""))
    cat("\\hline\n")

    temp=paste(" & \\multicolumn{",colno-1,"}{c}{",
               tnames[1],"} ",sep="")
    for(i in 2:tcount){
        #if(class(tnames[i])=="factor") temp=levels(tnames)[i]
        #else temp=tnames[i]
        temp=paste(temp," &  &\\multicolumn{",colno-1,"}{c}{",
                   tnames[i],"}",sep="")
    }
    temp=paste(temp," \\\\ \n",sep="")
    cat(temp)

    temp=paste("\\cline{2-",colno,"}",sep="")
    for(i in 2:tcount) {
        start=colno+2+(i-2)*(colno)
        temp=paste(temp,"\\cline{",start,"-",start+colno-2,"}",sep="")
    }
    cat(temp,"\n")
    temp=" "
    for(i in 1:tcount) {
        for(j in 1:(length(result[[i]]$cn))) {
            if(j==1) {
                if(i>1) temp=paste(temp," & ",sep="")
            }
            else temp=paste(temp," & ",result[[i]]$cn[j],sep="")
        }
    }
    cat(temp,"\\\\ \n")
    temp=" "
    for(i in 1:tcount) {
        for(j in 1:(length(result[[i]]$ncount))) {
            if(j==1) {
                if(i>1) temp=paste(temp," & ",sep="")
            }
            else temp=paste(temp,result[[i]]$ncount[j],sep=" & ")
        }
        for(k in 1:(colno-length(result[[1]]$ncount)))
            temp=paste(temp," & ",sep="")
    }
    cat(temp,"\\\\ \n")
    cat("\\hline\n")
    for(i in 1:dim(result[[1]]$out1)[1]){
        temp=""
        for(k in 1:tcount){
            if(k==1) {
                for(j in 1:colno) {
                    if(j==1) temp=r(result[[k]]$out1[i,j])
                    else temp=paste(temp,r(result[[k]]$out1[i,j]),sep=" & ")
                }
            }
            else {
                for(j in 1:colno){
                    if(j==1) temp=paste(temp," & ",sep="")
                    else temp=paste(temp,r(result[[k]]$out1[i,j]),sep=" & ")
                }
            }
        }
        cat(temp,"\\\\ \n")
    }
    cat("\\hline\n")
    if(caption.placement=="bottom")
        cat(paste("\\multicolumn{",tcn,"}{",
                  caption.position,"}{",caption,"}\\\\ \n",sep=""))
    cat("\\end{tabular}\n")
    cat(paste("\\end{",Fontsize[size],"}\n",sep=""))
    cat("\\end{table}\n")
}
